debugging-modules/README.md

```diff
@@ -0,0 +1,264 @@
+> GC safe points) to avoid making intermediate states or optimizations
+> observable. This can get a bit hairy, and whether we want to support this or
+> not is an open question.
+
+### Listing Sources and Displaying Source Text
+
+Debugger GUIs often display a list of source files or a file and directory tree
+of source files. To construct the complete list of source files for Wasm modules
+in a runtime, perform the following steps:
+
+* For each debuggee module in the runtime:
+    * Get or create the debugging module `SourceDebugger` for the debuggee
+      module.
+    * Append the result of calling `SourceDebugger.listSources` to the source
+      list.
+
+When a user selects a source from the list of all source files, a debugger GUI
+will typically display the selected source's text in a tab or panel. The text
+for a source can be retrieved by calling `SourceDebugger.getSourceText`.
+
+### Setting a Breakpoint
+
+When a user asks the debugger to set a source-level breakpoint, the debugger
+should perform the following steps:
+
+* Get the `SourceDebugger` associated with the breakpoint's source.
+* Let `options` be the `SourceBreakpointOptions` describing the requested
+  breakpoint.
+* Let `id` be the result of calling `SourceDebugger.setBreakpoint` with
+  `options`.
+    * If there is no expression at the breakpoint's location:
+        *  The `SourceDebugger.setBreakpoint` implementation should return null.
+    * Otherwise:
+        * The `SourceDebugger` implementation should set as many Wasm
+          breakpoints as necessary to pause just before evaluation of the
+          breakpoint location's expression begins using
+          `WasmDebugger.setBreakpoint`.
+* If `id` is null:
+    * Then no breakpoint was set (e.g because the requested location is within a
+      comment) and the UI should reflect this.
+* Otherwise:
+    * Save the `id` for if/when the breakpoint gets hit or the user wants to
+      clear the breakpoint.
+
+### Hitting a Breakpoint
+
+* Let `dbg` be the `SourceDebugger` instance for the current frame.
+* Construct the `WasmBreakInfo` for the Wasm breakpoint that was hit.
+* Let `breakResult` be the result of calling `dbg.onBreak` with the
+  `WasmBreakInfo`
+* If `breakResult.kind` is `"Continue"`:
+    * Resume the debuggee module's execution.
+* Otherwise if `breakResult.kind` is `"Pause"`:
+    * Assert that `breakResult.location` is not null.
+    * Update the debugger UI to show that execution is paused at the source
+      location `breakResult.location`.
+
+### Stepping
+
+* Let `dbg` be the `SourceDebugger` instance for the current frame.
+* Construct the `SourceStepOptions` for the step that the user requested.
+* Call `dbg.onStep` with the `SourceStepOptions`
+    * The `onStep` implementation should use the `WasmDebugger` to set Wasm
+      breakpoint(s) where it determines execution should pause after taking the
+      requested step.
```

Comment:
It is not possible to implement "step-over" just by setting breakpoints, because the debugger does not know what is the next source line that will be executed. It might not be the one directly succeeding the current line, as we could be in a loop, or some conditional construct.
Debuggers can examine what instructions are being executed and work out all of the possible branch targets, setting breakpoints on all of them. To do this, the debugger module should be able to disassemble the wasm module bytecode (which means that they also need access to that bytecode). 
A simpler (but very inefficient) alternative can be to send a number of instruction-level 'step' requests to the engine, until a different line is reached.